File: section2/106.从中序与后序遍历序列构造二叉树.py
#
# @lc app=leetcode.cn id=106 lang=python3
#
# [106] 从中序与后序遍历序列构造二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __str__(self):
        return f'node(val={self.val})'

    @staticmethod
    def generate(nums: list):
        root = TreeNode(nums[0])
        nodes = [root]

        index = 1
        while index < len(nums):
            new_nodes = []
            for node in nodes:
                if node:
                    node.left = None if nums[index] is None else TreeNode(nums[index])
                    node.right = None if nums[index + 1] is None else TreeNode(nums[index + 1])

                new_nodes.append(node.left if node else None)
                new_nodes.append(node.right if node else None)
                index += 2

            nodes = new_nodes

        return root

    def show(self):
        stacks = [self]
        nodes = []
        while stacks:
            row, new_stacks = [], []
            for node in stacks:
                nodes.append(node.val if node else None)
                if node:
                    new_stacks.append(node.left)
                    new_stacks.append(node.right)

            stacks = new_stacks

        return nodes


class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if len(inorder) == 0:
            return None

        node = TreeNode(postorder[-1])
        index = 0
        while index < len(inorder):
            if inorder[index] == postorder[-1]:
                break

            index += 1

        node.left = self.buildTree(inorder[:index], postorder[:index])
        node.right = self.buildTree(inorder[index + 1:], postorder[index:len(postorder) - 1])

        return node

    def test(self):
        print(self.buildTree([9, 3, 15, 20, 7], [9, 15, 7, 20, 3]).show())

Solution().test()
# @lc code=end

File: section2/section/106.从中序与后序遍历序列构造二叉树.go
package section

import "fmt"

/*
 * @lc app=leetcode.cn id=106 lang=golang
 *
 * [106] 从中序与后序遍历序列构造二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(inorder []int, postorder []int) *TreeNode {
	if len(postorder) == 0 {
		return nil
	}

	headValue := postorder[len(postorder)-1]
	splitIndex := 0
	for index, inorderValue := range inorder {
		if headValue == inorderValue {
			splitIndex = index
			break
		}
	}

	return &TreeNode{
		Val:   headValue,
		Left:  buildTree(inorder[0:splitIndex], postorder[0:splitIndex]),
		Right: buildTree(inorder[splitIndex+1:], postorder[splitIndex:len(postorder)-1]),
	}
}

func Test106() {
	// [3,9,20,null,null,15,7]
	fmt.Println(buildTree([]int{9, 3, 15, 20, 7}, []int{9, 15, 7, 20, 3}).ToValues())

	// [-1]
	fmt.Println(buildTree([]int{-1}, []int{-1}).ToValues())
}

// @lc code=end
